### DISCUSSION TITLE
migrate from mongoose to use typegoose with inconsistent data

### DISCUSSION THREAD

**Discussion by revati** [typegoose/typegoose#859] [Q&A] [answered]
Hello, i have existing system with messy and barely defined schemas with mongoose.

I tried to migrate one to typegoose - Users. Put explicit field types, added missing fields (that were in db but not defined in mongoose schemas). Everything is fine.

Everything worked fine, till one case in prod, where user cant update anything. When trying to update email, it says that name is required. And in typegoose scheme name is marked as required (it was in mongoose definition as well). But in db it is empty. (I believe reason is because all places that uses mongoose does not work with instances but update everything with `userModel.updateOne({_id: id}, { $set: { status } })` and it bypasses all validations. 

So code `newUserModel.find({_id: id}).changeEmail(newEmail).save()` does not work and throws error about name not being present. And those cases pops up like crazy. Every day there are few cases, and i have got into this endless loop, where i need to make every field optional, because there is some bad row, that has this inconsistency.

Is there better way how to allow to save email, but not trigger unrelated field validations? Or am i doomed?
I know i could too use `newUserModel.updateOne({_id: id}, {$set: {email: newEmail}})` but i don't like this approach, as it bypasses email validation, and logic required for email change (changing other fields as well).

Its bad that we have such messy data, but it is what it its and we will clean them up over time, but for now, we are sort of stuck in this limbo, where we cannot reliably migrate to typegoose because of those validation errors. I understand reasoning to validate entity completeness, but for our use case it would be so great to validate only fields that actually has been changed.

**Comment by revati** [comment]
I ubderstand that this issue is more related to mongoose it self and typegoose is just a wrapper.

**Comment by hasezoey** [comment, marked as answer]
> So code newUserModel.find({_id: id}).changeEmail(newEmail).save() does not work and throws error about name not being present

this is to be expected, mongoose by default validates *all* paths on `.save`, this can be changed with option [`validateModifiedOnly: true`](https://mongoosejs.com/docs/api/document.html#Document.prototype.save())

> I know i could too use newUserModel.updateOne({_id: id}, {$set: {email: newEmail}}) but i don't like this approach, as it bypasses email validation

`updateOne` supports to run validators on that path, but it is disabled by default, it can be enabled with option [`runValidators: true`](https://mongoosejs.com/docs/validation.html#update-validators)

**Reply by revati** [reply]
This is exactly it. I played around with it and found that it can even be set in global schema level:
```ts
// @ts-expect-error
@modelOptions({ schemaOptions: { validateModifiedOnly: true } })
```

Even tho it isnt explicitly allowed option, with @ts-expect-error it works as expected :)
